### DISCUSSION TITLE
Improved Branch Management and Issue Tracking on GitHub

### DISCUSSION THREAD

**Discussion by redblindbananaz** [The-Digital-Dilemma/GetIT#31] [General]
During my recent work on our project, I noticed that branches created directly from our local machines do not show up in our GitHub repository. This can potentially increase the risk of multiple team members unknowingly working on the same aspect of the project simultaneously. To address this issue and enhance our collaboration, I propose implementing the following practices:

1. Create an Issue or work from an existing issue
2. Branch from Established Issues: Instead of creating branches directly from our local machines, let's create branches from the established issues on GitHub. By doing this, everyone will be aware that a branch has been created and that work is being done on that particular aspect of the project.
3. Update Issue Status: To keep the team informed about the progress of our work, let's update the status of the issue to "In Progress" when we start working on it. This will provide visibility to the entire team that someone is actively working on that specific issue.

Please share your thoughts and feedback on this proposal.

**Comment by redblindbananaz** [comment]
@brilliant-light @makikamiya @charlie-canning-whitecliffe

**Reply by makikamiya** [reply]
Number 3 (Update Issue Status) makes sense to me; if I see an issue that has been assigned and is in progress, then I know that work is being done on that particular aspect of the project. I have to think more about numbers 1 and 2. Sometimes the purpose of creating a branch on the local machine is that we may want to test out an idea (which may or may not be related to a current issue). Sometimes the idea pans out and we may then decide to publish the branch for others to look at/test/etc. But often the idea doesn't work or you realise that it will take too long to implement, in which case the branch just remains on our local machine (for our own reference or maybe worked on at a later date). I'm reluctant to publish every branch that I might decide to create during a project.

**Reply by redblindbananaz** [reply]
I completely understand that for experimental features or testing purposes, working on a local branch would be the best approach. The suggestion to consider branching locally arose after we both worked on the same aspect of the project. Since I couldn't see any branches or issues in progress related to that work, I assumed it was available for me to work on. Currently, we are in the basic/MVP phase of the project, following a set of requirements, with minimal experimentation. In my opinion, once we move beyond the MVP phase, your methods will make more sense.

Another aspect to consider is that by keeping a branch locally, other team members won't have visibility into aspects of the work that could be beneficial for their own tasks or to review and test the progress.

I believe the most important step to take on, is to have issues status change to **In progress** allowing us to signal what we are working on to avoid doubling up( especially without issue assignment protocols) Then double up and branching local/hub would not be a problem with that status rule implementation and everyone can choose local vs remote 😀

**Reply by brilliant-light** [reply]
I think number 3 is also the best way to go, and using the issue status will still indicate that work is being done on a particular part of the project, which can also include testing something locally (if that makes sense) so I think it covers both of your points in a way

**Reply by makikamiya** [reply]
I completely agree with your thoughts that number 3 is the important one, @redblindbananaz, re issue-tracking and status updates as the mechanism by which we can see what everyone is working on and their progress. Moreover, I agree that seeing an unassigned issue, you're right to assume that it is not being worked on, and you can rightly pick it up.

While I understand your suggesting there may be beneficial reasons to publish local branches to the repo and that it's a choice to do so or not, we may have to agree to disagree on this, as I think there are compelling reasons against it:

- Having lots of branches in a repo makes it harder to understand the project's structure, especially when you have multiple branches that were created by multiple contributors but have never resolved (Have they been abandoned? Are they still work in-progress? Can I work on the branch, or has it been superseded and is no longer relevant to the project? Etc.)
- Local branches may contain unfinished work that is not ready for others to see or use. If they're published to the repo, I have no control over what someone else does to it, with likely headaches trying to resolve merge conflicts.
- Publishing all local branches blurs the distinction between stable, tested code and code that is experimental or in-progress. I believe a project's main repo should only contain code that is ready for others to use.
- There may be parts of the codebase that you don't want exposed publicly.
- Each open branch that is published increases the likelihood of merge conflicts and rebases.

For me, being fairly selective about which branches are published is also a workflow decision that results in a more manageable codebase.

**Reply by redblindbananaz** [reply]
I agree with you @makikamiya , we can find beneficial aspects in both methodologies and it seems that we are all agreeing on the importance of the status of an issue and the assignment to a team member to avoid duplicate work.
